File: snippets/fsharp/System/Type/BaseType/basetype3.fs
namespace global

// <Snippet2>
type A() = class end 

type B() = inherit A()

type C() = inherit B()   

module Example =
    [<EntryPoint>]
    let main _ =
        for t in typeof<A>.Assembly.GetTypes() do
            printfn $"{t.FullName} derived from: "
            let mutable derived = t
            while derived <> null do
                derived <- derived.BaseType
                if derived <> null then 
                    printfn $"   {derived.FullName}"
            printfn ""
        0
// The example displays the following output:
//       Example derived from:
//          System.Object
//       
//       A derived from:
//          System.Object
//       
//       B derived from:
//          A
//          System.Object
//       
//       C derived from:
//          B
//          A
//          System.Object
// </Snippet2>

File: snippets/csharp/System/Type/BaseType/basetype3.cs
// <Snippet2>
using System;

public class Example
{
    public static void Main()
    {
        foreach (var t in typeof(Example).Assembly.GetTypes())
        {
            Console.WriteLine($"{t.FullName} derived from: ");
            var derived = t;
            do
            {
                derived = derived.BaseType;
                if (derived != null)
                    Console.WriteLine($"   {derived.FullName}");
            } while (derived != null);
            Console.WriteLine();
        }
    }
}

public class A { }

public class B : A
{ }

public class C : B
{ }
// The example displays the following output:
//       Example derived from:
//          System.Object
//
//       A derived from:
//          System.Object
//
//       B derived from:
//          A
//          System.Object
//
//       C derived from:
//          B
//          A
//          System.Object
// </Snippet2>
